freqtrade/optimize/backtesting.py

```diff
@@ -218,6 +218,12 @@ def _get_close_rate(self, sell_row: Tuple, trade: LocalTrade, sell: SellCheckTup
         """
         # Special handling if high or low hit STOP_LOSS or ROI
         if sell.sell_type in (SellType.STOP_LOSS, SellType.TRAILING_STOP_LOSS):
+            if trade.stop_loss > sell_row[HIGH_IDX]:
+                # our stoploss was already higher than candle high,
+                # possibly due to a cancelled trade exit.
+                # sell at open price.
+                return sell_row[OPEN_IDX]
+
             # Set close_rate to stoploss
             return trade.stop_loss
```

Comment:
wouldn't a simple `min(trade.stop_loss, sell_row[HIGH_IDX])` be the better / simpler option (now if HIGH or OPEN is up for debate)?

---

Why would we use "open" if the price is above high - and not high?
You could also say we're using low (The most conservative) ... but what will be the most reliable value? (i really don't know!)

Reply:
> wouldn't a simple min(trade.stop_loss, sell_row[HIGH_IDX]) be the better / simpler option (now if HIGH or OPEN is up for debate)?

presumably we want to keep the original stoploss if it actually does occur during the candle that we're processing, even if it's higher than the open price?

> Why would we use "open" if the price is above high - and not high?
You could also say we're using low (The most conservative) ... but what will be the most reliable value? (i really don't know!)

My reasoning behind choosing "open" in this case is based on the assumption that since the stoploss was already higher than the candle high before this candle started, it should trigger at the "earliest" possible price.  I'm not entirely clear on the order of operations here, so this might not necessarily be accomplishing what I hoped.

Reply:
> My reasoning behind choosing "open" in this case is based on the assumption that since the stoploss was already higher than the candle high before this candle started, it should trigger at the "earliest" possible price.

This makes sense to me.

> I'm not entirely clear on the order of operations here, so this might not necessarily be accomplishing what I hoped.

freqtrade assumes open -> low -> high -> close.

Reply:
> freqtrade assumes open -> low -> high -> close.

that's not the full picture.
in general, freqtrade assumes this - however, stoploss "high" (trailing stop reference value) is assumed to happen first - there's no signal based on that - but the stoploss is adjusted first, then checked - which is the problem we're seeing with very tight stoplosses (which adjust to high - and then immediately sell as low is > high - stoploss distance).

The problem is - having it this way fixes the reverse problem - where price SHOULD be adjusted first (out of a static stoploss) ... so with this, it can also remain in the trade for longer (which can also be just as realistic, and is just as likely).